src/expire.c

```diff
@@ -507,10 +507,12 @@ void expireGenericCommand(client *c, long long basetime, int unit) {
 
     if (getLongLongFromObjectOrReply(c, param, &when, NULL) != C_OK)
         return;
-
     if (unit == UNIT_SECONDS) when *= 1000;
     when += basetime;
-
+    if (when <= 0 || (unit == UNIT_SECONDS && when >= LLONG_MAX / 1000)) {
+        addReplyErrorFormat(c, "invalid expire time in %s", c->cmd->name);
+        return;
+    }
```

Comment:
this check needs to move to before the adjustment above.

Reply:
Before the unit conversion, right?

Reply:
`when <= 0` makes EXPIRE to not take negative inputs at all.